Add initial porting guidelines

jashapiro opened the pull request:
Closes #71

This is the initial draft of the porting guidelines for add modules from `OpenScPCA-analysis` to `OpenScPCA-nf`. I tried to cover everything I had thought of in #71 and then some, but I am sure that what I have here is still incomplete! 

I also made a few changes to the workflow itself: 
- adding `emit:` statements to a couple workflows which I did mostly to test syntax (they have to be at the end, which is annoying, since that isn't the case for process outputs)
- removing a module-specific parameter (a container name) because apparently that is deprecated behavior (I have a question in the docs related to this)

I think the goal at this stage is to make sure we have most of the overall structure set, both from the POV of adding a modules and the structure of this document. I'm more confident in the former than the latter, though see some caveats below. I wrote the sections in kind of disconnected chunks (no IA!) , so it may be a bit choppy. Which is to say I am open to some significant restructuring!

Caveats:

- I assume a certain amount of familiarity with Nextflow, which I think is unavoidable, but I am happy to add more detail where it might be helpful. I also expect to add links to some relevant Nextflow docs, but I have not done that yet. 
- I don't yet have instructions about _using_ an emitted value as an input to another workflow; I think I wanted to wait until we had an example of that to really test it out

sjspielman commented on porting-modules.md:
```diff
@@ -0,0 +1,151 @@
+
+### Process granularity
+
+There are no hard and fast rules about how grandular each process should be, as we want to balance the workflow complexity with flexibility and runtime efficiency.
```
though i do very much like this word as a portmanteau of grand and granular, and i otherwise endorse it's use

sjspielman commented on porting-modules.md:
```diff
@@ -0,0 +1,151 @@
+The default resources for each process are 4 GB of memory and 1 CPU.
+Any additional requirements should be defined with `label` directives in the process definition.
+Available labels are defined in `config/process_base.config`, and separate labels are used for memory and CPU requirements.
+For example, to request 16 GB of memory and 4 CPUs, the process definition would include the following:
```
```suggestion

For example, to request 16 GB of memory and 4 CPUs, the process definition would include the following:
```

sjspielman commented on porting-modules.md:
````diff
@@ -0,0 +1,151 @@
+}
+```
+
+If an instance of a process fails, the memory requirements are automatically increased on the second and third attempts, but the general goal should be to have each process run to completion for the majority of samples with the assigned resources.
````
```suggestion
If an instance of a process fails, the memory requirements are automatically increased on the second and third attempts, but the general goal should be for each process successfully complete the majority of samples with the assigned resources.
```

sjspielman reviewed (commented):
Overall this seems like a good place to start from. I am sure as you are that content is missing, but there's so much here that I think it's a good baseline and we'll find out what is missing as more modules get ported...! The main missing items I noticed were (and these, certainly the 2nd, might be different PRs):

- I think we'd benefit from a section that describes the main workflow and its relationship to module more, including adding a module to the main workflow.
- This probably does not have a good one-size-fits-all answer, but -  how to actually _run_ the workflow, e.g. as part of development. 

Most of my suggestions focus on spacing and organization within the existing structure, but I think the overall order of topics here is fine at least for now.

jashapiro commented:
I did some reorganization and made updates and refinements that will hopefully address most of @sjspielman's comments. 

I added some comments about the main workflow (referred to here as the "default" workflow to distinguish it from the other entrypoints that do exist in the root `main.nf` file).
Container definitions are now moved to a central location.

I think I would leave the questions about running and testing to a future update, perhaps in concert with the next module being added. I have some thoughts, but I also want to take some more time to think about what to recommend and then we will have to test that those instructions are correct!

sjspielman reviewed (commented):
This is looking good! Content wise I don't have much to say at this point, other than for items which we plan to circle back to (like how to run/test during development), let's at least open a (small for now!) reminder issue to indeed circle back. The main reason I'm not approving now is b/c of the broken anchor link.

Another general comment I have is that I think we should do a round of copy editing here to update some passive voice -> active voice, [following guidelines we used for our docs](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#active-voice-and-present-tense) and use more "you" to refer to the readers here. 

For example, I might change this as follows:

```
# current sentence
Instead, the files that will be required for each sample should be selected using the `Utils.getLibraryFiles()` function or similar methods.

# you-ified
Instead, you should use the `Utils.getLibraryFiles()` function or similar methods to select files required for each sample.
```

I do think it's eminently reasonable for that to be a separate PR though, possibly bundled with other style decisions?

jashapiro commented on porting-modules.md:
```diff
@@ -0,0 +1,194 @@
+
+The function takes the following arguments:
+- `sample_dir` – The path to the sample directory
+- `format:` – The format of the files to be selected (`sce` or `anndata`)
+- `process_level:` – The processing level of the files to be selected (`raw`, `filtered` or `processed`)
```
The colons are actually an important part of the syntax here, so I would like to leave them.

sjspielman replied:
Ah, I see! At a quick glance I thought it was a typo with list explanations

jashapiro commented:
I made things much more (but not completely) active, and finished up the readme bullets that had escaped me earlier. 

I didn't use a lot of "you should" but instead made things more direct commandments, in a lot of cases. I hope that is okay.

sjspielman reviewed (approved):
LGTM! 🚀
